examples/measure/rs-measure.cpp

```diff
@@ -0,0 +1,553 @@
+    spat.set_option(RS2_OPTION_HOLES_FILL, 5); // 5 = fill all the zero pixels
+    // Define temporal filter
+    rs2::temporal_filter temp;
+    // Spatially align all streams to depth viewport
+    // We do this because:
+    //   a. Usually depth has wider FOV, and we only really need depth for this demo
+    //   b. We don't want to introduce new holes
+    rs2::align align_to(RS2_STREAM_DEPTH);
+
+    // Declare RealSense pipeline, encapsulating the actual device and sensors
+    rs2::pipeline pipe;
+
+    rs2::config cfg;
+    cfg.enable_stream(RS2_STREAM_DEPTH); // Enable default depth
+    // For the color stream, set format to RGBA
+    // To allow blending of the color frame on top of the depth frame
+    cfg.enable_stream(RS2_STREAM_COLOR, RS2_FORMAT_RGBA8);
+    auto profile = pipe.start(cfg);
+
+    auto sensor = profile.get_device().first<rs2::depth_sensor>();
+    
+    // TODO: At the moment the SDK does not offer a closed enum for D400 visual presets
+    // (because they keep changing)
+    // As a work-around we try to find the High-Density preset by name
+    // We do this to reduce the number of black pixels
+    // The hardware can perform hole-filling much better and much more power efficient then our software
+    auto range = sensor.get_option_range(RS2_OPTION_VISUAL_PRESET);
+    for (auto i = range.min; i < range.max; i += range.step)
+        if (std::string(sensor.get_option_value_description(RS2_OPTION_VISUAL_PRESET, i)) == "High Density")
+            sensor.set_option(RS2_OPTION_VISUAL_PRESET, i);
+
+    auto stream = profile.get_stream(RS2_STREAM_DEPTH).as<rs2::video_stream_profile>();
+    auto intrinsics = stream.get_intrinsics(); // Calibration data
+
+    // Create a simple OpenGL window for rendering:
+    window app(stream.width(), stream.height(), "RealSense Measure Example");
+
+    // Define application state and position the ruler buttons
+    state app_state;
+    app_state.ruler_start = { 0.45f, 0.5f };
+    app_state.ruler_end   = { 0.55f, 0.5f };
+    register_glfw_callbacks(app, app_state);
+
+    // After initial post-processing, frames will flow into this queue:
+    rs2::frame_queue postprocessed_frames;
+
+    // In addition, depth frames will also flow into this queue:
+    rs2::frame_queue pathfinding_queue; 
+
+    // Alive boolean will signal the worker threads to finish-up
+    std::atomic_bool alive{ true };
+
+    // The pathfinding thread will write its output to this memory:
+    std::vector<pixel> path;
+    float total_dist = 0.f;
+    std::mutex path_mutex; // It is protected by a mutex
+
+    // Video-processing thread will fetch frames from the camera,
+    // apply post-processing and send the result to the main thread for rendering
+    // It recieves synchronized (but not spatially aligned) pairs 
+    // and outputs synchronized and aligned pairs
+    std::thread video_processing_thread([&]() {
```

Comment:
IMHO The example should be refactored a bit and the threads should be encapsulated in objects of `PathFinder` and `VideoProcessor`. The classes will hold and manage the thread, and provide the necessary queues as getters/setters of the class.
This will make the `main` easier on the eye (which I think is required here) and also show high quality code standards as a C++ example

Reply:
What you say is better for long term re-usability and maintenance, but after implementing a quick draft I disagree it would benefit this demo. 
Current approach keeps many dependencies *implicit* (through `[&]` capture) creating more linear code-structure that is easier to read. Having many scattered classes and functions IMHO creates mental barrier that prevents readers from grasping the flow as whole.

Reply:
Fair enough